### PR TITLE
Reveal top x cards

### PR DESCRIPTION
**Pull request by poixen**
Updated reveal top card to reveal top x cards
Sometimes you want to reveal more than only the top 1 card.

max

### CODE REVIEW

**Comment by poixen** [comment]
@ZeldaZach this is a server change, you need to try it locally :P

**Comment by Daenyth** [comment]
Can we have reveal top cards to player? Like fateseal or jace, I don't get to see my own cards

**Comment by ZeldaZach** [comment]
Thanks for the notice poix, and I agree with dae

**Comment by poixen** [comment]
Yea should work fine. Give it a go with local 2 player.

**Comment by poixen** [comment]
Ah, seems you cant reveal cards in local games as its protected card info. Might need to be tested on the test server in that case.

**Comment by poixen** [comment]
Limited to the max number of cards in the deck, also added the limit to the popup.

**Comment by ZeldaZach** [comment]
Tested on the testing server and it seems to work as expected.

screenshot 2015-05-04 16 31 47

screenshot 2015-05-04 16 32 09
screenshot 2015-05-04 16 32 37

**Comment by tooomm** [comment]
There is a dot missing behind max: "Number of cards: (max**.** 53)"
Anyway, in favor of cards like [Traumatize](http://gatherer.wizards.com/Pages/Card/Details.aspx?multiverseid=370663) the wording
`Number of cards: (max %1)"` could be: `Number of cards: (%1 cards in library)`
Do you think that's any help for the user?

What happens if a user tries to put a higher value than his remaining cards?

<br>
@ZeldaZach did you use a compiled client with this commit? because your wording is old: card vs card**s** in the right click menu

**Comment by poixen** [comment]
@tooomm I have added the "." after max. The title of the window states the zone, I dont think we need it again.

**Comment by poixen** [comment]
@ZeldaZach tooomm has a point, you dont have the correct menu wording. Did you update your client?

**Comment by ctrlaltca** [comment]
:+1: 
One thing to note: old clients will still be able to use their "reveal one top card" option, so even back compatibility is ok.

**Comment by poixen** [comment]
Oh forgot to mention that, this doesn't replace functionality, just expands upon it. Thanks @ctrlaltca

**Comment by Daenyth** [comment]
If I use this feature in a game with a player who isn't updated will they
get a useful message? Or can I peek without them knowing?

File: common/server_player.cpp [hunk 1]
```diff
@@ -1522,7 +1522,14 @@ Response::ResponseCode Server_Player::cmdRevealCards(const Command_RevealCards &
         return Response::RespNameNotFound;
     
     QList<Server_Card *> cardsToReveal;
-    if (!cmd.has_card_id())
+    if (cmd.top_cards() != -1) {
+        for (int i = 0; i < cmd.top_cards(); i++) {
+            Server_Card *card = zone->getCard(i);
+            if (!card)
+                return Response::RespNameNotFound;
+            cardsToReveal.append(card);
```
Comment (Daenyth) [inline common/server_player.cpp]: Does cardsToReveal properly dispose of the pointer?

**Comment by ZeldaZach** [comment]
This doesn't seem to work right... when I type a number in, no matter the number, the entire library is revealed (to whomever you selected; i.e. If you select "all players" then everyone sees your library. If you select "zeldazach_t" then only zeldazach_t sees your entire library.)

screenshot 2015-05-07 09 58 27
screenshot 2015-05-07 09 58 40
screenshot 2015-05-07 09 58 42

File: common/server_player.cpp (hunk 1 above)
Comment (ctrlaltca) [inline common/server_player.cpp]: As far as i understand the code, no new object gets created here.

Comment (Daenyth) [inline common/server_player.cpp]: Derp you're right

**Comment by ctrlaltca** [comment]
@ZeldaZach can you argument on how this happened? i tested both 1,2 player local game and 2 players server game but i haven't been able to reproduce the issue.

**Comment by ctrlaltca** [comment]
> If I use this feature in a game with a player who isn't updated will they
> get a useful message? Or can I peek without them knowing?

You can't reveal your library only to yourself; you can reveal to all players or to a specific opponent.
This feature is leveraging on the old "reveal zone" method already existing on the server and known to clients, so even if you're using an old client you get the correct "Player reveals his library." message logged and you can actually see the revealed zone.

**Comment by ZeldaZach** [comment]
Here's how I tested and got my results;
ZeldaZach = build 5dfef62
ZeldaZach_old = build 07fcde0

What ZeldaZach does:
screenshot 2015-05-09 19 13 07
screenshot 2015-05-09 19 13 12

What ZeldaZach_old sees:
screenshot 2015-05-09 19 13 26

______
I then re-tested, using the "All players" option instead of just the opponent.

When ZeldaZach does:
screenshot 2015-05-09 19 15 14
screenshot 2015-05-09 19 15 21

What ZeldaZach sees:
screenshot 2015-05-09 19 15 26

What ZeldaZach_old sees:
screenshot 2015-05-09 19 15 29

_____

@ctrlaltca @poixen To me, it seems like backwards compatibility is nonexistent. If I'm doing something wrong, please feel free to correct me. I've left the steps here in order to help you guys reproduce what I'm experiencing.

**Comment by ctrlaltca** [comment]
So, i managed to reproduce @ZeldaZach 's problem.
If the server contains this patch, both old clients and new clients (with this patch) will get the feature working correctly.
Instead, if a "new" client connects to an "old" server, the "reveal" command works, but the server will ignore the number of cards to be revealed specified by the client, and will always reveal the whole deck.

**Comment by ctrlaltca** [comment]
As a workaround, a "new" client could set both the ```cmd.set_top_cards(number);``` and the ```cmd.set_card_id(0);``` fields, so that:
 * an old server will only understand the card_id field and reveal the first card;
 * a new server will parse the top_cards field first and correctly reveal the first X cards of the deck.

**Comment by poixen** [comment]
Merged @ctrlaltca fix. can be tested again.

**Comment by tooomm** [comment]
Just tested on the main server (1-player server game):
- `Reveal top cards to...` --> `All players` --> `7` (doesn't matter if i put 1, 7 or 20 there)<br>
Will result in the top card getting flipped on the library (same as `Always reveal top card`, but only once - not always), no library dialog with several cards
- Chat log will print `player reveals cardname from the top of his library.`

running newest weekly build (15dcfd1) on main server

**Comment by ZeldaZach** [comment]
1 player games are always iffy

**Comment by ctrlaltca** [comment]
The main server doesn't contain this patch, yet, so it will fallback to the old behavior.

**Comment by tooomm** [comment]
the old behavior doesn't open a window with just one card?
it just flips the *first* card on the library (exactly the same as `Always reveal top card`)?

**Comment by ctrlaltca** [comment]
Exactly

**Comment by tooomm** [comment]
ok.

@woogerboy21 you have scheduled an update already? ;)

**Pull request merged**